README.md
<img src="https://github.com/vishnumaiea/ptScheduler/blob/main/assets/ptScheduler-Feature-Image-1_3-1.png" alt="ptScheduler" width="500"/>

# Pretty tiny Scheduler
**Pretty tiny Scheduler** or **ptScheduler** is a non-preemptive task scheduler and timing library for **Arduino**. It helps you write **non-blocking, periodic tasks** easily without using ordinary delay routines or using `millis()` or `micros()` functions. Delay routines can prevent other parts of your code from running until the delay is exhausted. Using `millis()` function in your code can make it harder to understand and manage. ptScheduler solves these problems.

ptScheduler acts as a wrapper for `millis()` or `micros()` based logic used for timing of tasks. Under the hood, ptScheduler uses the native `micros()` implementation of Arduino. The `micros()` function is a hardware timer-based **ISR** that increments a global counter variable (unsigned integer) every microsecond. This function is available on all microcontrollers that support the Arduino framework. ptScheduler was using `millis()` function previously, but the new `micros()` provides granular timing in microseconds compared to milliseconds.

When you create a new ptScheduler task, you can specify the **time intervals** and **execution modes**. There are two execution modes - **ONESHOT** and **SPANNING**. Oneshot task is executed only once every time an interval is elapsed. Spanning task on the other hand remains active during the span of an interval.

All class member variables and functions are public and therefore give you full control over your tasks, allowing dynamically changing the behavior of the task.
To run a task, just enclose the `call()` function inside any conditional statements, either inside your infinite loop or inside a function. Every time you invoke the `call()` function, it checks if the elapsed time is larger than the preset interval. If yes, it will return `true` and cause the code under the conditional block to be executed.

ptScheduler is good mainly for **control applications** that require periodic polling of sensors, GPIOs, and other IO devices. ptScheduler tasks can coexist with preemptive tasks such as **FreeRTOS** tasks.

## Hello World

Here is the basic `Hello World` example.

```cpp
#include "ptScheduler.h"

// Create tasks
ptScheduler sayHello = ptScheduler (PT_TIME_1S); //time in microseconds

// Setup function runs once
void setup() {
  Serial.begin (9600);
}

// Infinite loop
void loop() {
  if (sayHello.call()) {  // Executed every second
    Serial.println ("Hello World");
  }

  // Add other tasks and non-blocking code here
}
```

## Tutorial

Complete tutorial can be found at **[CIRCUITSTATE](https://circuitstate.com/tutorials/ptscheduler-a-minimal-cooperative-task-scheduler-for-arduino/)**

## Documentation

See the reference of all available functions and variables in the **[Documentation](https://vishnumaiea.gitbook.io/ptscheduler/)** page.


Changes.md

#
**+05:30 09:52:45 AM 27-06-2023, Tuesday**

  * Added new functions - `isEnabled()` and `isSuspended()`.
  * 🆕 New version **2.2.0** 🎉
#
**+05:30 04:16:06 PM 29-03-2023, Wednesday**

  * Updated docs.
  * 🆕 New version **2.1.3** 🎉

#
**+05:30 03:57:49 PM 29-03-2023, Wednesday**

  * Added Doxygen-style descriptions.
  * Added more comments.
  * 🆕 New version **2.1.2** 🎉

#
**+05:30 08:09:52 PM 03-12-2022, Saturday**

  * Added new variable `toClearExecutionCounter` to control whether or not to reset the `executionCounter` when a task is suspended automatically or manually. Both behaviors may be useful in some cases. The default value is `false`.

#
**+05:30 10:43:12 PM 10-08-2022, Wednesday**

  * 🆕 New version **2.1.1**

#

**+05:30 10:24:50 PM 02-07-2022, Saturday**

  There was a bug 🪲 at the `setSequenceRepetition()` function. When the `sequenceRepetition` was 1, the `sequenceRepetitionExtended` would become 0 and thus a repeated task would run continuously. It was fixed by adding a condition to check if the `sequenceRepetition` is 1. Also I was using `sequenceLength` for the conditional check where it should have been odd checking on `sequenceRepetition` instead. Both are fixed now. FIXED ✅ [Issue #5](https://github.com/vishnumaiea/ptScheduler/issues/5) .

  ```cpp
  if (sequenceRepetition == 1) {
    sequenceRepetitionExtended = 1;
  }
  else if ((sequenceRepetition % 2) == 1) {
    sequenceRepetitionExtended = uint32_t(((sequenceLength + 1) / 2) * sequenceRepetition);
    sequenceRepetitionExtended--; //we need one less, otherwise there will be one extra active state
  }
  else {
    sequenceRepetitionExtended = uint32_t((sequenceLength * sequenceRepetition) / 2);
  }
  ```
#

**+05:30 11:07:41 AM 29-03-2022, Tuesday**

  I have renamed a few of the variables to better match their job.

  There is also going to be changes in the type of tasks. We will only have **ONESHOT** or **SPANNING** tasks. Everything else about a task can be customized.


#

**+05:30 05:32:14 PM 27-09-2021, Monday**

  `taskMode` has been renamed to `executionMode` because it makes more sense what the parameter controls.


#

**+05:30 05:04:34 PM 31-08-2021, Tuesday**

  I am going to rename Periodic tasks as "Infinite" and Iterated tasks as "Finite". Because both Periodic and Iterated are confusing and unintuitive. With these renaming, it will now be more clear. They can be described as "Finite Repetition Mode" and "Infinite Repetition Mode".

  Terms Oneshot and Spanning will be kept, since they are already meaningful.

  Now there can only be two type of tasks; Oneshot and Spanning with two modes; Finite and Infinite.

  Timing modes = Oneshot and Spanning
  Repetition modes = Finite and Infinite

  Each task can be given with an interval sequence. This is a set of intervals that is always greater than 0.

#

**+05:30 06:19:25 PM 14-06-2021, Monday**

  Let's commit this version as **Version 2.0.0** since this will be incompatible with the previous version.

#

**+05:30 03:50:27 PM 14-06-2021, Monday**

  Added `millis()` overflow workarounds. This is accomplished with a new function `timeElapsed()`. I also replaced all `millis()` instances with `micros()`.

  It is working now, but for some reason the `Hello-World.ino` prints "Hello World" twice every second! 😅

  Okay I found it. It was necessary to reset `prevTimeDelta` along with `elapsedTime` at the point of entry. Now it works flawlessly 🤍

#

**+05:30 11:59:55 AM 13-06-2021, Sunday**

  Ready for a next release with important bug fixes. **Version 1.1.2** 🎉

#

**+05:30 03:42:00 PM 11-06-2021, Friday**

  When does a task start? - A task starts when the first `call()` is made 📋

  When a task is resumed, should it respect the skip parameters? Currently, this is not respected. But what if I want to have that feature? One method is to disable the task and later enable it, instead of suspending and resuming. But this also causes every counter to freeze and thus you won't be able to read the sleep interval counters. You will have to manually and arbitrarily enable the task. But this assures that the skip parameters are respected.

  Another method is to keep the task sleep mode to `PT_SLEEP_SUSPEND` and call `resume()` to the resume the task. But just before calling `resume()`, set `taskStarted` state variable to `false`. This is because it is the `taskStarted` flag that determines if an interval has to be skipped. This flag is set to `true` when you invoke `call()` for the first time (after a duration that is greater than the set skip interval).

  So when we do the following for example, the skip intervals are respected.

  ```cpp
    eioTask.taskStarted = false;
    eioTask.resume();
  ```
#

**+05:30 09:46:25 PM 10-06-2021, Thursday**

  Found a bug 🐞 in the `oneshot()` and `spanning()` routines (lines 261 and 364). When checking for skip parameters, the conditional statement had two instances of `skipIntervalSet`. One should have been `skipIterationSet`. Due to this, `setSkipIteration()` was not working for iterated tasks. FIXED ✅ [Issue #2](https://github.com/vishnumaiea/ptScheduler/issues/2#issue-919728782)

#

**+05:30 04:16:52 PM 10-06-2021, Thursday**

  I plan to optimize the interface of `ptScheduler` by removing the redundant modes. There only needs to be four things to consider when defining a task; oneshot or spanning, and periodic or iterated. That yields 4 modes. Every other modes can be implemented using these four base properties. TODO ⌛

  Second improvement is to make `ptScheduler` use **microsecond timing**. This is already implemented in Arduino platform through `micros()` routine which should be based on a timer interrupt. This will require a new type to be defined, `time_us_t` which will still be a `uint64_t`. TODO ⌛


src/ptScheduler_VMA.h

//=======================================================================//

// --- Pretty tiny Scheduler ---

// Pretty tiny Scheduler is a small library for writing non-blockcing
// periodic tasks for Arduino without using traditional NOP delay routines.

// Author : Vishnu Mohanan (@vishnumaiea)
// Version : 0.1
// License : MIT
// Repo : https://github.com/vishnumaiea/ptScheduler

// Last modified : +05:30 10:29:34 PM 13-11-2020, Friday

//=======================================================================//
//includes

#include <Arduino.h>

//=======================================================================//
//main class

class ptScheduler {
  private :
    
  public :
    int64_t entry = 0;  //the entry point of a task, returned by millis()
    int64_t exit = 0; //the exit point of a tast, returned by millis()
    int64_t elapsed = 0;  //elapsed time since the last task execution
    int64_t residue = 0;
    int64_t interval = 0; //the interval at which the task will be called once
    uint64_t intervalCounter = 0; //how many intervals has been passed
    uint64_t taskCounter = 0; //how many times the task has been executed
    bool activated = true;  //whether a task is allowed to run or not
    bool started = false; //whether a task has started an execution cycle
    bool suspended = false; //whether a task is prevented from running until further activation
    bool ended = true;  //end of an execution cycle
    bool running = false; //a task is running
    
    ptScheduler(int64_t timeValue); //sets the initial interval for the task
    ~ptScheduler();
    void activate();  //activating a tast run at each intervals
    void deactivate();  //prevent a task from running at each intervals
    bool call();  //the task invokation, determines whether a task should be run or not
    void setInterval(int64_t timeValue);  //dynamically set task interval
};

//=======================================================================//


src/ptScheduler_VMA.cpp

//=======================================================================//

// --- Pretty tiny Scheduler ---

// Pretty tiny Scheduler is a small library for writing non-blocking
// periodic tasks for Arduino without using traditional NOP delay routines.

// Author : Vishnu Mohanan (@vishnumaiea)
// Version : 0.1
// License : MIT
// Repo : https://github.com/vishnumaiea/ptScheduler

// Last modified : +05:30 10:48:05 PM 13-11-2020, Friday

//=======================================================================//
//includes

#include "ptScheduler_VMA.h"

//=======================================================================//
//constructor

ptScheduler::ptScheduler(int64_t timeValue = 0) {
  interval = timeValue;
  activated = true;
}

//=======================================================================//
//destructor

ptScheduler::~ptScheduler() {
  
}

//=======================================================================//
//determines whether a task has to be executed.
//enclose this in a conditional statement to run that block of code.
//do not use other blocking delays.

bool ptScheduler::call() {
  if ((!started)) {
    if (interval < 0) {
      interval = abs(interval);
      entry = millis();
      started = true;
      running = false;
      suspended = true;
      ended = false;
      return false;
    }
    else {
      entry = millis();
      started = true;
      ended = false;
      intervalCounter++;

      if (activated) {
        running = true;
        suspended = false;
        taskCounter++;
        return true;
      }
      else {
        running = false;
        suspended = true;
        return false;
      }
    }
  }

  else if ((started)) {
    elapsed = millis() - entry;
    if (elapsed >= interval) {
      started = false;
      running = false;
      suspended = false;
      ended = true;
      exit = entry + elapsed;
      return false;
    }
    else {
      // started = true;
      // ended = false;
      running = false;
      suspended = true;
      return false;
    }
  }
  
  return false;
}

//=======================================================================//
//activates a task.
//only active tasks are executed regardless of their right to execution.

void ptScheduler::activate() {
  activated = true;
}

//=======================================================================//
//deactivates a task.

void ptScheduler::deactivate() {
  activated = false;
}

//=======================================================================//
//set the interval at which you want the task to be executed.
//time is in milliseconds.
//a positive value will execute the task at 0th time and then every time
//the set interval has passed. a minus value will postpone the task until
//the set interval has passed for the first time. after, the task is
//executed at every intervals.

void ptScheduler::setInterval(int64_t timeValue) {
  interval = timeValue;
}

//=======================================================================//


src/ptScheduler.h

//==============================================================================//
/**
 * @file ptScheduler.h
 * @author Vishnu Mohanan (@vishnumaiea)
 * @brief "Pretty tiny Scheduler" is an Arduino library for writing non-blocking
 * periodic tasks without using delay() or millis() routines.
 * 
 * Read the header file for a better understanding of how this library is
 * organized. I have added plenty of comments to help you with it.
 * 
 * @version 2.2.0
 * @link https://github.com/vishnumaiea/ptScheduler
 * @date Last modified : +05:30 09:53:34 AM 27-06-2023, Tuesday
 * @copyright License: MIT
 * 
 */
//==============================================================================//
// Includes

#include <Arduino.h>

//==============================================================================//
// Defines

// Change this to your desired serial port for printing the debug messages
#define debugSerial         Serial

#define  PT_MODE_ONESHOT    1
#define  PT_MODE_SPANNING   2

#define  PT_SLEEP_DISABLE   1    //self-disable mode
#define  PT_SLEEP_SUSPEND   2    //self-suspend mode

#define  PT_MS_MULTIPLIER   1000   //multiplier to convert ms to us

// Time periods
#define  PT_TIME_1MS        1000
#define  PT_TIME_2MS        2000
#define  PT_TIME_5MS        5000
#define  PT_TIME_10MS       10000
#define  PT_TIME_20MS       20000
#define  PT_TIME_50MS       50000
#define  PT_TIME_100MS      100000
#define  PT_TIME_200MS      200000
#define  PT_TIME_500MS      500000
#define  PT_TIME_1S         1000000
#define  PT_TIME_2S         2000000
#define  PT_TIME_3S         3000000
#define  PT_TIME_4S         4000000
#define  PT_TIME_5S         5000000
#define  PT_TIME_6S         6000000
#define  PT_TIME_7S         7000000
#define  PT_TIME_8S         8000000
#define  PT_TIME_9S         9000000
#define  PT_TIME_10S        10000000
#define  PT_TIME_1MIN       60000000  // This line was predicted by AI

// Frequenceis in terms of time values
#define  PT_FREQ_1HZ        PT_TIME_1S
#define  PT_FREQ_2HZ        PT_TIME_500MS
#define  PT_FREQ_5HZ        PT_TIME_200MS
#define  PT_FREQ_10HZ       PT_TIME_100MS
#define  PT_FREQ_20HZ       PT_TIME_50MS
#define  PT_FREQ_50HZ       PT_TIME_20MS
#define  PT_FREQ_100HZ      PT_TIME_10MS
#define  PT_FREQ_200HZ      PT_TIME_5MS
#define  PT_FREQ_500HZ      PT_TIME_2MS
#define  PT_FREQ_1KHZ       PT_TIME_1MS

#define  PT_TIME_DEFAULT    PT_TIME_1S

typedef uint64_t time_ms_t;  // Time in milliseconds
typedef uint64_t time_us_t;  // Time in microseconds

// Add your own timing functions here
#define  GET_MICROS         micros
#define  GET_MILLIS         millis

//==============================================================================//
//main class

class ptScheduler {
  private :
    
  public :
    time_us_t entryTime = 0;  // The entry time of a task, returned by micros()
    time_us_t exitTime = 0; // The exit time of a task, returned by micros()
    time_us_t elapsedTime = 0;  // Elapsed time since entry time
    time_us_t lastElapsedTime = 0;
    uint64_t intervalCounter = 0; // How many intervals have been passed
    uint64_t suspendedIntervalCounter = 0; // How many intervals have been passed after suspending the task
    uint64_t executionCounter = 0; // How many times the task has returned true
    uint64_t sequenceRepetitionCounter = 0; // How many times the sequence has been repeated

    uint32_t sequenceRepetition = 0;  // How many times an interval sequence has to be executed
    uint32_t sequenceRepetitionExtended = 0;  // Repetitions * interval sequence length
    uint8_t taskMode = PT_MODE_ONESHOT;  // The execution mode of the task
    uint8_t sleepMode = PT_SLEEP_DISABLE; // Default is disable
    uint8_t sequenceLength;  // How many intervals in a sequence
    time_us_t* sequenceList;  // A pointer to the interval sequence
    uint8_t sequenceIndex = 0;  // Index position of interval sequence
    uint32_t skipInterval = 0;  // Number of individual intervals to skip
    uint32_t skipSequence = 0; // Number of sequences (set of intervals) to skip
    time_us_t skipTime = 0; // Time to wait before running a task
    uint32_t prevTimeDelta = 0; // Previous time difference
    uint32_t timeDelta = 0; // Current time difference
    uint32_t microsValue = 0; // Value returned by micros()

    bool taskEnabled = true;  // Task is allowed to run or not
    bool taskStarted = false; // Task has started an execution cycle
    bool cycleStarted = false; // Task has started an interval cycle
    bool taskSuspended = false; // A task is prevented from running until further activation
    bool sequenceRepetitionEnded = false;  // End of a repetition
    bool taskRunning = false; // The task is running
    bool taskRunState;  // The current execution state of a task

    bool inputError = false;  // If any user input parameters are wrong
    bool skipIntervalSet = false; // If skip interval was set
    bool skipSequenceSet = false;  // If skip sequence was set
    bool skipTimeSet = false; // If skip time was set
    bool toClearExecutionCounter = false; // If the execution counter has to be cleared

    // Description of all functions can be found in the .cpp file
    ptScheduler (time_us_t interval_1);
    ptScheduler (uint8_t _mode, time_us_t interval_1);
    ptScheduler (uint8_t _mode, time_us_t* listPtr, uint8_t listLength);
    ~ptScheduler();
    void reset();
    void enable();
    bool isEnabled();
    void disable();
    void suspend();
    bool isSuspended();
    void resume();
    bool oneshot();
    bool spanning();
    bool call();
    bool setInterval (time_us_t value);
    bool setSequenceRepetition (int32_t value);
    bool setSkipInterval (uint32_t value);
    bool setSkipSequence (uint32_t value);
    bool setSkipTime (time_us_t value);
    bool setTaskMode (uint8_t mode);
    bool setSleepMode (uint8_t mode);
    bool isInputError();
    void printStats();
    void getTimeElapsed();
};

//==============================================================================//
